python/ray/tune/tests/test_tune_save_restore.py

```diff
@@ -0,0 +1,160 @@
+
+    def tearDown(self):
+        ray.shutdown()
+
+        # Without this line, test_tune_server.testAddTrial would fail.
+        _register_all()
+
+    def _get_trial_dir(self, absoulte_exp_dir):
+        trial_dirname = next(
+            (child_dir for child_dir in os.listdir(absoulte_exp_dir)
+             if (os.path.isdir(os.path.join(absoulte_exp_dir, child_dir))
+                 and child_dir.startswith(self.MockTrainable._name))))
+
+        trial_absolute_dir = os.path.join(absoulte_exp_dir, trial_dirname)
+
+        return trial_dirname, trial_absolute_dir
+
+    def _train(self, exp_name, local_dir, absolute_local_dir):
+        shutil.rmtree(absolute_local_dir, ignore_errors=True)
+        trial, = tune.run(
+            self.MockTrainable,
+            name=exp_name,
+            stop={"training_iteration": 1},
+            checkpoint_freq=1,
+            local_dir=local_dir,
+            config={
+                "env": "CartPole-v0",
+                "log_level": "DEBUG"
+            })
+
+        exp_dir = os.path.join(absolute_local_dir, exp_name)
+        _, abs_trial_dir = self._get_trial_dir(exp_dir)
+
+        self.assertIsNone(trial.error_file)
+        self.assertEqual(trial.local_dir, exp_dir)
+        self.assertEqual(trial.logdir, abs_trial_dir)
+
+        self.assertTrue(os.path.isdir(absolute_local_dir), absolute_local_dir)
+        self.assertTrue(os.path.isdir(exp_dir))
+        self.assertTrue(os.path.isdir(abs_trial_dir))
+        self.assertTrue(
+            os.path.isfile(
+                os.path.join(abs_trial_dir, "checkpoint_1/checkpoint-1")))
+
+    def _restore(self, exp_name, local_dir, absolute_local_dir):
+        trial_name, abs_trial_dir = self._get_trial_dir(
+            os.path.join(absolute_local_dir, exp_name))
+
+        checkpoint_path = os.path.join(
+            local_dir, exp_name, trial_name,
+            "checkpoint_1/checkpoint-1")  # Relative checkpoint path
+
+        # The file tune would find. The absolute checkpoint path.
+        tune_find_file = os.path.abspath(os.path.expanduser(checkpoint_path))
+        self.assertTrue(
+            os.path.isfile(tune_find_file),
+            "{} is not exist!".format(tune_find_file))
+
+        trial, = tune.run(
+            self.MockTrainable,
+            name=exp_name,
+            stop={"training_iteration": 2},  # train one more iteration.
+            restore=checkpoint_path,  # Restore the checkpoint
+            config={
+                "env": "CartPole-v0",
+                "log_level": "DEBUG"
+            })
+        self.assertIsNone(trial.error_file)
+
+    def testDottedRelativePath(self):
+        local_dir = "./test_dotted_relative_local_dir"
+        exp_name = self.prefix + "DottedRelativeLocalDir"
+        absolute_local_dir = os.path.abspath(local_dir)
+        self._train(exp_name, local_dir, absolute_local_dir)
+        self._restore(exp_name, local_dir, absolute_local_dir)
```

Comment:
These used to have `os.chdir(...)`. Are those not needed anymore?

Reply:
After rebasing the Ray, running the `tune.run()` no longer changes the current working directory. Printing out the `os.getcwd()` before and after `self._train` show no difference at my Mac, so I believe some commit yesterday fixed the problem though I don't know which it is.

Intuitively, running `tune.run()` should not change the working directory. So removing `os.chdir` is what I expect.

Reply:
Ah, that sounds great much better. Thanks for clarifying!